### PR TITLE
Added build & push to GH container registry & DockerHub

### PR DESCRIPTION
**Pull request by KoalaGeo**
# Overview

Same as https://github.com/geopython/pycsw/pull/714 

I've added a GitHub workflow which builds a container and pushes this to the GitHub container registry. This is a useful parallel container delivery system to Dockerhub,

I've also updated the installation.rst and running-with-docker.rst to reference this option - docker pull ghcr.io/geopython/pygeoapi:latest (this will only work after PR)

I didn't remove the dockerhub related sections, but after a test period the ghcr.io/geopython/pygeoapi could be added to all examples, and potentially the Dockerhub workflow deprecated.

See & test container - https://github.com/KoalaGeo/pygeoapi/pkgs/container/pygeoapi

This does also associate packages with the whole organisation so they're visible at https://github.com/KoalaGeo?tab=packages or they would be at https://github.com/geopython?tab=packages

### CODE REVIEW

**Comment by tomkralidis** [comment]
Any update on this issue?

**Comment by tomkralidis** [comment]
Thanks @justb4.

@justb4 @francbartoli @pvgenuchten @jorgejesus @kalxas @GeoSander what are your thoughts here on pygeoapi's Docker image publication:

1. GHCR & DockerHub
2. GHCR Only
3. DockerHub only
4. Other options?

**Comment by KoalaGeo** [comment]
@tomkralidis Sorry, Xmas and COVID working it's way through the family. 

I was thinking from @justb4 review the workflow would build both dockerhub & ghcr, but the docs would refer to both existing but give examples re dockerhub to keep them readable/avoid duplication. 

Then orgs that have issues hitting dockerhub limits could use ghcr. 

Happy to write whatever the team decide

**Comment by justb4** [comment]
To clarify from my review: what I think is important is the following workflow-step order for the Docker Image:

* build the Docker Image ONLY ONCE
* run the unit tests with this Image
* publish IMage to either GHCR and/or DockerHub

Only if first two steps are successful Docker push is started. In case of 2 Docker target repo's, doing separate builds/pushes (YAML files) can create inconsistencies between Docker images (think of a GH commit/push while one is building).

Under these conditions (steps) I am in favour of option 1 (GHCR and DockerHub).

**Comment by francbartoli** [comment]
+1 for option one

**Comment by tomkralidis** [comment]
@KoalaGeo let's go with option 1 with the workflow proposed by @justb4.

**Comment by justb4** [comment]
Keep up the good work @KoalaGeo ! Hope we all can help you. Tough material.

Had an interesting find trying to run unit tests with Docker Container: #848 . Running tests with 100% success would be required within the `containers.yml` workflow as a step between build and push. For some tests the issue is more general as I see many commented out in [main.yml](https://github.com/geopython/pygeoapi/blob/master/.github/workflows/main.yml#L77) workflow..

**Comment by KoalaGeo** [comment]
Getting there, using https://github.com/docker/metadata-action which wasn't available before to add tags to the images as it should be a neater solution than previous . 

However having issue getting it to pick up tags, when a release is made

**Comment by webb-ben** [comment]
Realizing https://github.com/geopython/pygeoapi/pull/866 is a duplicate of this. Just want to make whatever platform the pygeoapi image is published on offers both ARM64 and AMD64 architectures :)

**Comment by justb4** [comment]
I just revisited and this PR looks quite complete, it addresses the three conditions I earlier outlined:

- only run when `main.yml` (main Build) successful
- build for both DockerHub and GHCR
- push only when tests run with Docker passing

Though my knowledge of GHCR build/push is lacking.
And the ARM64 and AMD64 architectures may be missing, see #866.

Proof will be in the pudding, once we run as `geopython/pygeoapi`...

File: docs/source/installation.rst [hunk 1]
```diff
@@ -48,11 +48,23 @@ pip
...
    docker pull geopython/pygeoapi:latest
+   
+Using GitHub Container Registry   
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`Docker image <https://github.com/geopython/pygeoapi/pkgs/container/pygeoapi>`_
+
+.. code-block:: bash
+
+   docker pull ghcr.io/geophython/pygeoapi:latest
```
[the remark this answers is not in the document]
Comment (KoalaGeo) [reply docs/source/installation.rst]: Fixed

**Comment by KoalaGeo** [comment]
Had a look again and this should be doing what you'd like @justb4 

It runs only when Build is Complete
Think I've got 1 build
Then it runs the tests
Then it pushes to docjkerhub & the GHCR. 

If you accept the PR need to set:
username: ${{ secrets.DOCKER_USERNAME }}
password: ${{ secrets.DOCKER_PASSWORD }}

File: .github/workflows/containers.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          tags: |
+            type=semver,pattern={{version}}
+            type=raw,value=latest,enable=${{ endsWith(github.ref, github.event.repository.default_branch) }}
+      
+      - name: Build and push Docker images
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
```
[the remark this answers is not in the document]
Comment (KoalaGeo) [reply .github/workflows/containers.yml:74]: Accepted

**Review by justb4 — APPROVED** [review]
Approving. The proof is really when this is within the repo as this can't be tested in other contexts. So let's merge and work from there. Already thanks @KoalaGeo !

**Pull request merged**